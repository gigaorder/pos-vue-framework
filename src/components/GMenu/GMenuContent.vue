<script>
  import {computed, onBeforeUnmount, onMounted, ref, watch} from '@vue/composition-api';
  import menuable from '../../mixins/menuable';
  import getVModel from '../../mixins/getVModel';
  import {convertToUnit} from '../../utils/helpers';
  import detachable from '../../mixins/detachable';
  import ClickOutside from '../../directives/click-outside/click-outside';
  import Resize from '../../directives/resize/resize';
  import stackable from '../../mixins/stackable';
  import dependent from '../../mixins/dependent';

  export default {
    name: 'GMenuContent',
    directives: {
      ClickOutside,
      Resize
    },
    props: {
      activator: null, //ref HTMLElement
      // basic
      ...{
        value: Boolean,
      },
      // positioning
      ...{
        auto: Boolean,
        top: Boolean,
        bottom: Boolean,
        left: Boolean,
        right: Boolean,
        offsetX: {
          type: Boolean,
        },
        offsetY: {
          type: Boolean,
          default: true
        },
        nudgeLeft: {
          type: [Number, String],
          default: 0
        },
        nudgeRight: {
          type: [Number, String],
          default: 0
        },
        nudgeTop: {
          type: [Number, String],
          default: 0
        },
        nudgeBottom: {
          type: [Number, String],
          default: 0
        },
        positionX: [Number, String],
        positionY: [Number, String],
        allowOverflow: Boolean,
        offsetOverflow: Boolean,
        absolute: Boolean,
        clientX: [Number, String],
        clientY: [Number, String],
      },
      // toggling
      ...{
        openOnHover: Boolean,
        closeOnClick: {
          type: Boolean,
          default: true
        },
        closeOnContentClick: Boolean
      },
      // sizing
      ...{
        maxWidth: {
          type: [Number, String],
          default: '100%'
        },
        minWidth: [Number, String],
        maxHeight: {
          type: [Number, String],
          default: 'auto'
        },
        minHeight: [Number, String],
        contentFillWidth: {
          type: Boolean,
          default: true
        }
      },
      // delay
      ...{
        openDelay: {
          type: [Number, String],
          default: 0
        },
        closeDelay: {
          type: [Number, String],
          default: 0
        }
      },
      // dependent mixin
      ...{
        closeDependents: {
          type: Boolean,
          default: true,
        },
        isDependent: {
          type: Boolean,
          default: true,
        },
      },
      //styling content
      ... {
        contentClass: String
      }
    },
    setup(props, context) {
      const isActive = getVModel(props, context);
      const {attachToRoot, detach} = detachable(props, context);
      const {
        updateDimensions, dimensions, computedTop, computedLeft, calcXOverflow, calcYOverFlow
      } = menuable(props, context);
      const {getMaxZIndex} = stackable(props, context)


      function getResizeObserver() {
        let activatorResizeObserver = undefined

        const init = () => {
          activatorResizeObserver = new ResizeObserver(() => {
            context.root.$nextTick(() => {
              updateDimensions(props.activator.value)
            })
          })
          props.activator.value && activatorResizeObserver.observe(props.activator.value)
        }

        const destroy = () => {
          if (activatorResizeObserver) {
            activatorResizeObserver.disconnect()
            activatorResizeObserver = undefined
          }
        }

        return {
          observer: activatorResizeObserver, init, destroy
        }
      }

      const resizeObserver = getResizeObserver(props)

      // init resize observer when activator is mounted to parent
      watch(props.activator, () => {
        if (!resizeObserver.observer) resizeObserver.init()
      })

      // update dimensions when toggled on
      watch(() => props.value, newVal => {
        if (newVal) {
          context.root.$nextTick(() => {
            updateDimensions(props.activator.value)
          })
        }
      }, {lazy: true})

      let rootEl
      const getOpenDependentElements = ref(null)

      onMounted(function () {
        rootEl = attachToRoot()
        getOpenDependentElements.value = dependent(this)
      })

      onBeforeUnmount(() => {
        resizeObserver.destroy();
        if (rootEl) rootEl.removeChild(context.refs.content) // menu content is mounted to root element
      })

      const calculatedLeft = computed(() => {
        const menuWidth = Math.max(dimensions.content.width, parseFloat(calculatedMinWidth.value))
        return convertToUnit(calcXOverflow(computedLeft.value, menuWidth)) || '0'
      })
      const calculatedTop = computed(() => {
        return convertToUnit(calcYOverFlow(computedTop.value)) || '0'
      })
      const calculatedMaxWidth = computed(() => {
        return convertToUnit(props.maxWidth) || '0'
      })
      const calculatedMinWidth = computed(() => {
        if (props.contentFillWidth) return convertToUnit(Math.max(dimensions.activator.width, dimensions.content.width))
        if (props.minWidth) return convertToUnit(props.minWidth) || '0'

        const minWidth = Math.min(dimensions.content.width, state.pageWidth);
        const _calculatedMaxWidth = isNaN(calculatedMaxWidth.value) ? minWidth : parseInt(calculatedMaxWidth.value)
        return convertToUnit(Math.min(_calculatedMaxWidth, minWidth)) || 0;
      })
      const calculatedMaxHeight = computed(() => {
        return convertToUnit(props.maxHeight) || '0';
      })

      const contentStyles = computed(() => ({
        top: calculatedTop.value,
        left: calculatedLeft.value,
        maxHeight: calculatedMaxHeight.value,
        minWidth: calculatedMinWidth.value,
        maxWidth: calculatedMaxWidth.value,
        zIndex: isActive.value ? getMaxZIndex() + 2 : 6
      }))

      const contentListeners = {
        ...props.closeOnContentClick && {
          click() {
            if (isActive.value) isActive.value = false
          }
        },
        ...props.openOnHover && {
          mouseleave() {
            if (isActive.value) isActive.value = false
          }
        }
      }

      const genDirectives = () => {
        //callback to close menu when clicked outside
        const closeConditional = (e) => {
          const target = e.target;
          return isActive.value && context.refs.content && !context.refs.content.contains(target)
        }
        const clickOutsideDirective = {
          name: 'click-outside',
          value: () => {
            isActive.value = false
          },
          arg: {
            closeConditional: closeConditional,
            include: () => {
              return [context.parent.$el, ...getOpenDependentElements.value()];
            }
          },
        }
        //equates to v-show="value" in template
        const vShowDirective = {
          name: 'show',
          value: props.value
        }
        const resizeDirective = {
          name: 'resize',
          value: () => updateDimensions(props.activator.value)
        }

        const directives = [vShowDirective, resizeDirective]
        if (!props.openOnHover && props.closeOnClick) directives.push(clickOutsideDirective)
        return directives;
      }

      const contentClasses = computed(() => ({
        'g-menu--content': true,
        [props.contentClass]: !!props.contentClass
      }))

      const genMenuContent = () => <div style={contentStyles.value}
                                        class={contentClasses.value}
                                        ref="content"
                                        {...{directives: genDirectives(), on: contentListeners}}>
        {context.slots.default && context.slots.default()}
      </div>;

      return {
        isActive,
        getOpenDependentElements,
        genMenuContent
      }
    },
    render() {
      return this.genMenuContent()
    }
  }
</script>

<style scoped>

</style>
