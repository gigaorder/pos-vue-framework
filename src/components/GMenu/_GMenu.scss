.menu {
  display: inline;
  position: relative;

  &-content {
    z-index: $z-index-menu;
    position: absolute;
    display: inline-block;
    max-width: 80%;
    overflow-y: auto;
    overflow-x: hidden;
    contain: content;
    @include elevation(2);
  }
}
