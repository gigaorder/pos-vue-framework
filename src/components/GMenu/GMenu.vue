<script>
  import {getInternalValue} from '../../mixins/getVModel';
  import { onBeforeUnmount, onMounted, reactive, ref, watch } from 'vue';
  import ClickOutside from '../../directives/click-outside/click-outside';
  import delayable from '../../mixins/delayable';
  import GMenuContent from './GMenuContent';
  import { Fragment } from 'vue-fragment'
  import { getScopeIdRender } from '../../utils/helpers';

  export default {
    name: 'GMenu',
    components: { GMenuContent, Fragment },
    directives: {
      ClickOutside
    },
    props: {
      // basic
      ...{
        modelValue: Boolean,
        eager: Boolean,
        activator: null,
      },
      // positioning
      ...{
        //TODO: prop for menu to appear in center
        auto: Boolean,
        top: Boolean,
        bottom: Boolean,
        left: Boolean,
        right: Boolean,
        offsetX: {
          type: Boolean,
        },
        offsetY: {
          type: Boolean,
          default: true
        },
        nudgeLeft: {
          type: [Number, String],
          default: 0
        },
        nudgeRight: {
          type: [Number, String],
          default: 0
        },
        nudgeTop: {
          type: [Number, String],
          default: 0
        },
        nudgeBottom: {
          type: [Number, String],
          default: 0
        },
        positionX: [Number, String],
        positionY: [Number, String],
        allowOverflow: Boolean,
        offsetOverflow: Boolean,
        absolute: Boolean,
      },
      // toggling
      ...{
        openOnHover: Boolean,
        closeOnClick: {
          type: Boolean,
          default: true
        },
        closeOnContentClick: Boolean
      },
      // sizing
      ...{
        maxWidth: {
          type: [Number, String],
          default: '100%'
        },
        minWidth: [Number, String],
        maxHeight: {
          type: [Number, String],
          default: 'auto'
        },
        minHeight: [Number, String],
        contentFillWidth: Boolean
      },
      // delay
      ...{
        openDelay: {
          type: [Number, String],
          default: 0
        },
        closeDelay: {
          type: [Number, String],
          default: 0
        }
      },
      //styling content
      ...{
        contentClass: String,
      }
    },
    setup(props, context) {
      const isActive = getInternalValue(props, context);
      const { runDelay } = delayable(props)
      const content = ref(null)
      const activatorEl = ref(null);
      let activatorVNode

      const state = reactive({
        top: 0,
        hasJustFocused: false,
        isFirstRender: true,
        clientX: 0,
        clientY: 0,
      });

      onMounted(function () {
        activatorEl.value = activatorVNode && activatorVNode[0] && activatorVNode[0].el
      })

      onBeforeUnmount(() => isActive.value = false)

      function toggleContent(event) {
        const activator = event.target || event.currentTarget;
        if (!activator) {
          return
        }
        if(props.absolute) {
          state.clientX = event.clientX
          state.clientY = event.clientY
        }
        isActive.value = !isActive.value;
      }

      watch(isActive, val => {
        if (val && !props.eager && state.isFirstRender) state.isFirstRender = false;
      }, { lazy: true })

      const genContent = () => {
        const contentOptions = {
          ...{
            // props
            ...Object.assign({}, props, {
              modelValue: isActive.value
            }),
            activator: activatorEl,
            hasJustFocused: state.hasJustFocused,
            clientX: state.clientX,
            clientY: state.clientY
          },
          'onUpdate:modelValue': (e) => {
              isActive.value = e
              state.hasJustFocused = e
          },
          ref: content
        }
        return <g-menu-content {...contentOptions}>
          {context.slots.default()}
        </g-menu-content>
      }

      const on = !props.disabled && {
        ...props.openOnHover && {
          mouseenter(event) {
            runDelay('open', () => {
              if (state.hasJustFocused || isActive.value) {
                return
              }
              toggleContent(event);
              state.hasJustFocused = true;
            })
          },
          mouseleave(event) {
            runDelay('close', () => {
              if (content && content.value.$el.contains(event.relatedTarget)) {
                return
              }
              isActive.value = false
              state.hasJustFocused = false
            })
          }
        },
        click(event) {
          toggleContent(event)
        }
      }

      const genWrapper = function () {
        const _activatorVNode = context.slots.activator && context.slots.activator({ toggleContent, on });
        activatorVNode = _activatorVNode

        return <div>
          {_activatorVNode}
          {props.eager ? genContent() : (!state.isFirstRender && genContent())}
        </div>;
      }

      return {
        state,
        isActive,
        genWrapper,
        activatorEl
      }
    },
    render() {
      return getScopeIdRender()(this.genWrapper.bind(this))()
    }
  }
</script>

<style scoped lang="scss">
  @import "../../style/elevation";

  .g-menu--content {
    position: absolute;
    display: inline-block;
    overflow-y: auto;
    overflow-x: hidden;
    contain: content;
    @include elevation(2);
  }
</style>
