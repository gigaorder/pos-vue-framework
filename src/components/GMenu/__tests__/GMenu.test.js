import Vue from 'vue/dist/vue.common.js'
import VueCompositionApi from '@vue/composition-api'
// init plugin first
Vue.use(VueCompositionApi)

const createElement = require('../../../tests/setup').createElement
// import component here
import GMenu from '../GMenu';

describe('Menu', () => {
  it('should render default and activator slots', () => {
    const vm = new Vue({
      template: `
        <div data-app>
          <g-menu v-model="showMenu">
            <template v-slot:activator="{toggleContent}">
              <button @click="toggleContent">Activator</button>
            </template>
            <div>content</div>
          </g-menu>
        </div>
      `,
      components: { GMenu },
      data() {
        return {
          showMenu: false
        }
      }
    }).$mount(createElement());
    expect(vm.$children[0].$scopedSlots.default).toBeTruthy()
    expect(vm.$children[0].$scopedSlots.activator).toBeTruthy()
  })

  //fixme: test fails when run as file
  it('should attach content to root dom', function () {
    const vm = new Vue({
      template: `
        <div data-app>
          <div>
            <g-menu v-model="showMenu">
              <template v-slot:activator="{toggleContent}">
                <button @click="toggleContent">Activator</button>
              </template>
              <div>content</div>
            </g-menu>
          </div>
        </div>
      `,
      components: { GMenu },
      data() {
        return {
          showMenu: false
        }
      }
    }).$mount(createElement());
    expect(vm.$el.firstChild.className).toBe('menu-content')
  });

  it('should not mount content if lazy loading', function () {
    const vm = new Vue({
      template: `
        <div data-app>
          <g-menu v-model="showMenu" lazy>
            <template v-slot:activator="{toggleContent}">
              <button @click="toggleContent" width="100" height="50">Activator</button>
            </template>
            <div>content</div>
          </g-menu>
        </div>
      `,
      components: { GMenu },
      data() {
        return {
          showMenu: false
        }
      }
    }).$mount(createElement());
    expect(vm.$el.querySelector('.menu-content')).toBe(null)
  });

  it('should activate on click', function () {
    const vm = new Vue({
      template: `
        <div>
          <div data-app>
            <g-menu v-model="showMenu">
              <template v-slot:activator="{toggleContent}">
                <button @click="toggleContent">Activator</button>
              </template>
              <div>content</div>
            </g-menu>
          </div>
        </div>
      `,
      components: { GMenu },
      data() {
        return {
          showMenu: false
        }
      }
    }).$mount(createElement());
    vm.$nextTick(() => {
      // expect(vm.$el.outerHTML).toMatchSnapshot()
      const button = vm.$el.querySelector('button')
      button.click()
      expect(vm.showMenu).toBe(true)
    })
  });
})
