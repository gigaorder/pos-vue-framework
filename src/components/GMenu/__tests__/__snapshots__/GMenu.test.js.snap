// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should activate on click 1`] = `"<div><div data-app=\\"\\"><div class=\\"menu-content\\" style=\\"top: 0px; left: 0px; max-height: auto; min-width: 0px; max-width: 80%; display: none;\\"> <div>content</div></div><div class=\\"menu\\"><div><button>Activator</button></div></div></div></div>"`;

exports[`Menu should attach content to root dom 1`] = `"<div data-app=\\"\\"><div class=\\"menu-content\\" style=\\"top: 0px; left: 0px; max-height: auto; min-width: 0px; max-width: 80%; display: none;\\"> <div>content</div></div><div><div class=\\"menu\\"><div><button>Activator</button></div></div></div></div>"`;

exports[`Menu should not mount content if lazy loading 1`] = `"<div data-app=\\"\\"><div class=\\"menu\\"><div><button width=\\"100\\" height=\\"50\\">Activator</button></div></div></div>"`;

exports[`Menu should render default and activator slots 1`] = `"<div data-app=\\"\\"><div class=\\"menu-content\\" style=\\"top: 0px; left: 0px; max-height: auto; min-width: 0px; max-width: 80%; display: none;\\"> <div>content</div></div><div class=\\"menu\\"><div><button>Activator</button></div></div></div>"`;
